### PR TITLE
Add ELBO.

### PR DESCRIPTION
**Pull request by taku-y**
I added a function which calculates ELBO (Evidence Lower Bound). The convergence was checked on advi.ipynb.

### CODE REVIEW

**Comment by springcoil** [comment]
Greate work :) I'll add this here http://arxiv.org/pdf/1506.03431.pdf as a link just for our own documentation. (And my own knowledge)
I reviewed the code and I have no issues with it (I also think it helps me understand some of my own problems with theano scan).

It's been tested on Advi - but may be a good idea to add some sort of test for this. This means that any of the commits for variational inference in the future can be trusted better. 

@twiecki @fonnesbeck and @jsalvatier anything to add?

File: pymc3/variational/ADVI.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 import theano
 from ..theanof import make_shared_replacements, join_nonshared_inputs, CallableTensor, gradient
 from theano.tensor import exp, concatenate, dvector
+import theano.tensor as t
```
Comment (twiecki) [inline pymc3/variational/ADVI.py]: We try to move to importing as `T`.

File: pymc3/variational/ADVI.py [hunk 2]
```diff
@@ -98,6 +103,22 @@ def inner_gradients(logp, n, uw):
...
     return duw
 
+def inner_elbo(logp, n, uw):
+    # uw contains both, mean and diagonals
+    l = (uw.size/2).astype('int64')
+    u = uw[:l]
+    w = uw[l:]
+
+    r = MRG_RandomStreams(seed=1)
+    n = r.normal(size=(10, u.tag.test_value.shape[0]))
```
Comment (twiecki) [inline pymc3/variational/ADVI.py]: Drawing 10 MC samples?

**Comment by twiecki** [comment]
This is great. I have two questions: don't we already compute the elbo (just not store it), why does it require a second function?

Does this slow down ADVI?

**Comment by taku-y** [comment]
@springcoil Thank you for your comment and reviewing the code. 

I agree that adding some tests is better for future updates. I'm going to learn how to write tests for PyMC3.

**Comment by taku-y** [comment]
@twiecki I think you might want to accurate estimates of ELBO for monitoring the optimization process, for example, 1 ELBO calculation per 100 parameter updates (not supported the current code). For this purpose, I wrote a separate function. But as you said we already compute ELBO, so I will use it for computational efficiency. 

For reference, I put two plots of the learning curve (ELBO) with n_mcsamples_elbo=1, 50. We can see the curve become smooth, although fluctuated due to the randomness of the gradient. 

n_mcsamples1

n_mcsamples50

**Comment by fonnesbeck** [comment]
I agree, this is useful for diagnostics, and we should probably add a `plot_elbo` function.

Are we happy merging this (particularly since it is just to another feature branch)?

**Comment by springcoil** [comment]
I'm happy to merge this after a rebase!

**Comment by fonnesbeck** [comment]
Will rebase to master.

**Comment by twiecki** [comment]
@taku-y I understand your code better now and added the accurate elbo estimate back in.

**Comment by taku-y** [comment]
Thanks! :)

**Pull request merged**